text/0005-is-operator.md

```diff
@@ -0,0 +1,163 @@
+
+```
+permit(
+  principal is User,
+  action == Action::"viewFile",
+  resource is File in Folder::"public"
+);
+```
+
+At evaluation time, `principal is User` will evaluate to true if the type of `principal` is `User`, and `resource is File in Folder::"public"` will evaluate to true if the type of `resource` is `File` and it is a member of `Folder::"public"`.
+
+## Motivation
+
+The `is` operator allows users to restrict the scope of their policies to particular entity types.
+
+In many cases, this can also be done by adding restrictions to the schema.
+For example, the `appliesTo` field for each action in the schema allows users to specify which entity types can be used as principals and resources for that action.
+In the example [above](#basic-example), the schema could specify that `applyFile` should only apply to principals of type `User` and resources of type `File`.
+However, this doesn't help if the user is not using a schema, the action is unconstrained, or the policy refers to an action that may act on multiple principal/resource types.
+
+For example, here is one case where the action is unconstrained, taken from the Cedar [document cloud example](https://github.com/cedar-policy/cedar-examples/tree/main/cedar-example-use-cases/document_cloud).
+
+```
+forbid (principal, action, resource)
+when
+{
+  resource has owner &&
+  principal != resource.owner &&
+  resource has isPrivate &&
+  resource.isPrivate
+};
+```
+
+This policy applies to any principal, action, resource pair, but requires that the resource has `owner` and `isPrivate` attributes.
+If the resource does not have these attributes, then the policy is a no-op.
+
+Say that only one entity type (e.g., `File`) has the fields `owner` and `isPrivate`.
+Then this policy can be written more succinctly as follows.
+
+```
+forbid (principal, action, resource is File)
+when
+{
+  principal != resource.owner &&
+  resource.isPrivate
+};
+```
+
+## Detailed design
+
+### Changes required to parser/AST
+
+- Update the grammar and AST to support expressions of the form `e is et` where `et` is a `Name` (e.g., `Namespace::EntityType`).
+- Update the grammar and AST to support expressions of the form `e1 is et in e2` (but not `e1 in e2 is et`, because this is more difficult to mentally parse).
+- Update the grammar to allow expressions like `e is et` and `e1 is et in e2` in the policy scope for principals and resources.
+The action part of the scope will only allow expressions like `action == Action::"foo"` and `action in [Action::"foo", Action::"bar"]`, as before.
+
+### Changes required to evaluator
+
+`e is et` will evaluate to true iff `et` is a `Name` (e.g., `Namespace::EntityType`) and `e` is an expression that is an entity of type `et`.
+Otherwise it will evaluate to false.
+To maintain the property that "conditions in the scope never error," `e is et` will never return an error; it will simply return false if any error occurs.
```

Comment:
What possible errors are there? The only one I can think of is that `e` evaluates to an entity reference that does not exist in the store. If that's the only error, let's call it out specifically -- it's the same semantics as for `e1 in e2` when `e1` doesn't exist, right?

Reply:
Evaluating `e is et` shouldn't even require an entity store lookup

Reply:
I think @mwhicks1 insinuated a look up upon the schema definition, not the "entity store"

Reply:
Evaluation never looks at the schema, and I presume no one here is proposing to change that?

Reply:
Evaluating `e is et` only requires looking at the literal typename component of `e`'s UID

Reply:
In the policy scope, we only allow `principal is ...` and `action is ...`, so failure will never occur. But you could still have ill-typed expression in the policy condition. Looking again at the implementation of evaluation for `in`, it looks like we do want to report type errors here rather than just evaluating to `false`.

So, proposed revised text:
```
`e is et` will evaluate to true iff `et` is a `Name` (e.g., `Namespace::EntityType`) and `e` is an expression that is an entity of type `et`.
If `e` is an entity-typed expression, but does not have type `et`, then `e is et` will evaluate to false.
If `e` is a non-entity-typed expression, the evaluator will produce a type error.

Note that this design maintain the property that "conditions in the scope never error" because `principal` and `resource` are guaranteed to be entities, so the expression `(principal|resource) is et` will never error.
```

Reply:
like `2 is number`, where we'd somehow have to distinguish the builtin type `number` from the entity type `number`

Reply:
> Evaluating e is et can only result in an error if e is a non-entity-typed expression (e.g., "foo" is User). 

Of course! Not sure what I was thinking ...